### ISSUE TITLE
Implement mini-batch gradient descent, adaptive learning rate, and GPU acceleration support

### ISSUE THREAD

**Issue by AmirhosseinHonardoust** [guilhermedom/multilayer-perceptron#1]
**Description:**
the current multilayer perceptron implementation appears to use full-batch gradient descent and runs exclusively on cpu, which limits scalability and slows training on larger datasets. additionally, the learning rate is fixed throughout training. introducing mini-batch training, adaptive learning-rate scheduling, and optional gpu acceleration (via `torch`, `tensorflow`, or `cupy`) would significantly improve convergence speed and numerical efficiency.

**Steps to reproduce:**
1. run the training script (`multilayer_perceptron.py`) on a medium-sized dataset ( >10 000 samples).
2. observe that all samples are used per epoch and no batching is applied.
3. monitor training speed and resource utilization, cpu usage is high, gpu remains idle.
4. check the code, there is no implementation of batch iteration or learning-rate scheduling.

**Expected behavior:**
* implement **mini-batch gradient descent** with configurable batch size (`--batch_size 32`).
* integrate **adaptive learning-rate schedulers**, such as time-based decay or optimizers like `Adam`, `RMSProp`, or `Adagrad`.
* optionally detect cuda devices and use **gpu acceleration** if available:

  ```python
  device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
  model.to(device)
  ```
* print average loss per mini-batch and overall epoch metrics for better monitoring.
* ensure results are reproducible via fixed random seeds (`np.random.seed()` or framework-specific seeds).

**Actual behavior:**
* training uses full dataset per epoch; no batching implemented.
* fixed learning rate leads to slow or unstable convergence depending on dataset size.
* gpu resources (if present) are unused, resulting in long training times on larger problems.

**Environment:**
* OS: Windows 10
* Python version: 3.10
* Branch: main

**Additional context:**
adding batch-based training reduces gradient noise and increases performance, while adaptive learning rates stabilize convergence across different datasets. gpu support will also allow users to experiment with deeper architectures and larger data. consider refactoring the training loop into a class (`MLPTrainer`) that encapsulates optimizer, device, and batch management for cleaner extensibility.

**Comment by henriquepgomide** [comment]
@AmirhosseinHonardoust , hold your bots or change the prompts;

Even Microsoft does not support Win10 😂
